### DISCUSSION TITLE
Support revision  #233

### DISCUSSION THREAD

**Discussion by wancom** [wonder-wonder/cakemix-server#236] [Design]
As explained #233, some essential feature for revision is not provided.
In next version, we will implement for revision support features.

**Comment by wancom** [comment]
My current idea is as follows.

Add new APIs.
- For getting old revision data
  `GET /doc/:docid/rev/:revision`
  docid is terget document uuid and revison is the revision to get.
- For reverting the document to old revision
  `PUT /doc/:docid/revert/:revision`
  docid is terget document uuid and revision is the revision to revert

Revert function involves OT session stopping to keep consistency.
Revert function may not be necessary.

**Reply by 0eta0** [reply]
how does the revert function work? Is it just replaced a revision document with the latest one?

**Reply by wancom** [reply]
Yes, that's right.
The function works like copy and paste.

**Reply by 0eta0** [reply]
Thanks, I got it.

**Comment by wancom** [comment]
@0eta0 Any ideas?

**Reply by 0eta0** [reply]
Is there any function to get a difference of code?

**Reply by wancom** [reply]
No, there isn't now.
But it's good idea. We need more discussion about this one.

**Comment by wancom** [comment]
New idea (getting diff) given by @0eta0 

For getting diff with current and old version.
- GET /doc/:docid/diff/:revision/:revisionbase
docid is terget document uuid.
revison is the revision to compare.
revisonbase is the revision compared. (If omitted, the current version is used.)

**Reply by wancom** [reply]
Implemention hint

https://qiita.com/shibukawa/items/dd75ad01e623c4c1166b

**Reply by wancom** [reply]
I want an idea for data structure.
Can you show me the diff data structure which can handle easily on front?

**Reply by wancom** [reply]
@0eta0

**Reply by 0eta0** [reply]
Codemirror supports diff viewer, I'm going to use it. 
https://codemirror.net/demo/merge.html

**Reply by 0eta0** [reply]
this function will work on the front side.
https://github.com/codemirror/CodeMirror/blob/master/demo/merge.html
